### ISSUE TITLE
Any suggestions for regularization?

### ISSUE THREAD

**Issue by drscotthawley** [state-spaces/mamba#88]
Dear Mamba-SSM team, congratulations on your success!  Obviously many of us are excited about exploring the applications of your work. 

Since there's no dropout in your model, what do you suggest for imposing regularizations?  I applied the [drop-in Mamba replacement](https://www.reddit.com/r/MachineLearning/comments/18d65bz/d_thoughts_on_mamba/) that was used with Karpathy's GPT example to my [MIDI Transformer Colab](https://colab.research.google.com/drive/1SFYwzCGYYpJHjp6JX1kPamD8g0cq6lPM#scrollTo=6TugZVD0pIER).  (which is paired with a pre-Mamba [blog post](https://drscotthawley.github.io/blog/posts/Transformers2-MusicBox.html)). 

Compared to the original multi-head attention version, the Mamba-powered version runs faster, but also overfits a lot more.  Meaning that losses on the validation set [bottom out sooner but *higher*](https://wandb.ai/drscotthawley/musicbox-jsb-tutorial?workspace=user-drscotthawley) with Mamba than with the vanilla multi-head attention.   

 So far, I've tried 

* increasing weight decay
* clipping gradient values

...trying values expressed in your paper.  But nothing I've tried seems to have any regularizing effect. 

What model parameters would you suggest for improving generalization? 

Thanks!

**Comment by chazzmoney** [comment]
I’m very interested to see the response here from the authors, but I’ll say something potentially wrong.

i’m not familiar with the architecture used to create MIDI using the transformer mechanism, but in the original MAMBA paper they were pretty clear that continuous signals do not benefit form MAMBA, and that something like S4 seemed to outperform. So the performance you may be seeing may have less to do with regularization and more to do with the application of MAMBA to this problem.

**Comment by tridao** [comment]
You can use dropout, just like Transformers. It's not implemented here but you can add it.

**Comment by ElliottDyson** [comment]
> You can use dropout, just like Transformers. It's not implemented here but you can add it.

This thread may be more useful if tagged as a feature request then? (Just for traceability to mitigate it from being lost)

**Comment by Anri-Lombard** [comment]
This is still an issue; I trained the model on 10 epochs for a specific use case and the loss just decreases drastically to overfit quickly 
![Uploading Screenshot 2024-07-06 at 15.52.09.png…]()

**Comment by Anri-Lombard** [comment]
Has anyone implemented this with some success?

**Comment by windsornguyen** [comment]
I find that adding dropout decreases performance for state space models. Does anyone else also observe this phenomenon?

**Comment by Anri-Lombard** [comment]
I added dropout and it actually improved performance for me.
![training_convergence](https://github.com/user-attachments/assets/a8c1bc7a-0161-41f3-b798-dcc91075143d)

**Comment by Anri-Lombard** [comment]
Although MAMBA still does very bad in my use case, it just learned better.
